### PR TITLE
FieldConfig: support overrides model

### PR DESCRIPTION
**Pull request by ryantxu**
This PR modifies how we store field overrides so that they are applied one property at a time.

This is the first step towards making general per/field configuration style.

In a future PR, this should be extended to:
* incorporate `replaceVariables` function in the override process
* auto set the min/max values across all numeric fields
* more `fieldDisplay` & `displayValue` cleanup -- we are approaching a much more reusable approach
* Make sure the same processed DataFrames[] can be used in the inspector -- and my extension CSV downloader!

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
This looks very solid to me. 

With a unit test for the main applyFieldOverrides function it could be merged I think.

**Comment by ryantxu** [comment]
now with a basic unit test :)   will add some more with the auto-min/max PR that will follow

**Review by torkelo — APPROVED** [review]

File: packages/grafana-data/src/field/fieldOverrides.test.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    // Hardcode the max value
+    f0.fields[1].config.max = 0;
+    f0.fields[1].config.decimals = 6;
+
+    const src: FieldConfigSource = {
+      defaults: {
+        unit: 'xyz',
+        decimals: 2,
+      },
+      overrides: [
```
Comment (dprokop) [inline packages/grafana-data/src/field/fieldOverrides.test.ts]: Really like how you made the overrides work with Matchers

File: packages/grafana-data/src/field/fieldOverrides.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            });
+          }
+        }
+      }
+
+      // Set the Min/Max value automatically
+      if (field.type === FieldType.number) {
+        if (!isNumber(config.min) || !isNumber(config.max)) {
+          if (!range) {
+            range = findNumericFieldMinMax(data);
```
Comment (dprokop) [inline packages/grafana-data/src/field/fieldOverrides.ts]: Do i understand correctly than min/max is set as min/max across entire dataframe? Why is that, shouldn't it be as field's min/max?

File: packages/grafana-data/src/field/fieldOverrides.ts [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+  } else if (input) {
+    // skips empty string
+    if (key === 'unit' && input === 'none') {
+      return null;
+    }
+  }
+  return input;
+}
+
+export function setDynamicConfigValue(config: FieldConfig, options: DynamicConfigValueOptions) {
```
Comment (dprokop) [inline packages/grafana-data/src/field/fieldOverrides.ts]: Not sure we need config as an argument here as it's passed via `options` too. And mutating the config feelt unintuitive to me at first sight. But this is just a nit.

File: packages/grafana-data/src/field/fieldOverrides.ts [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+      const config: FieldConfig = { ...field.config } || {};
+      if (field.type === FieldType.number) {
+        setFieldConfigDefaults(config, source.defaults);
+      }
+
+      // Find any matching rules and then override
+      for (const rule of override) {
+        if (rule.match(field)) {
+          for (const prop of rule.properties) {
+            setDynamicConfigValue(config, {
```
Comment (dprokop) [inline packages/grafana-data/src/field/fieldOverrides.ts]: Regarding the other comment about mutating the config- I feel I see your point now. Makes this code easier without the need for reassignments

File: packages/grafana-data/src/field/fieldOverrides.ts (hunk 2 above)
Comment (ryantxu) [reply packages/grafana-data/src/field/fieldOverrides.ts]: It is a good question, and I *think* this is the right default.  When looking at a graph of a bunch of signals, the graph auto min-max to the range of everything, not each line.

I think we can (and should!) have an override option that sets the min|max to the individual field

File: packages/grafana-data/src/field/fieldOverrides.ts (hunk 3 above)
Comment (ryantxu) [reply packages/grafana-data/src/field/fieldOverrides.ts]: ya -- I went back and forth on mutating vs always setting one property.  I can go back, but let me explain what I was thinking:

In the beginning of the `applyFieldOverrides` loop, the config is copied and then all the rules are attached to it -- inside a few more loops.

The setXXX functions are really internal to this file -- they are only exported so the tests can see them.  Note they are not exported in the `index.ts` file, only:

```
export { applyFieldOverrides } from './fieldOverrides';
```

File: packages/grafana-data/src/field/fieldOverrides.ts (hunk 4 above)
Comment (ryantxu) [reply packages/grafana-data/src/field/fieldOverrides.ts]: yes -- reassignments got pretty ugly

File: packages/grafana-data/src/field/fieldOverrides.ts (hunk 3 above)
Comment (dprokop) [reply packages/grafana-data/src/field/fieldOverrides.ts]: @ryantxu yup, got it when reading the `applyFieldOverrides` fn:)

**Pull request merged**